File: auxiliary/Thrift/src/vib/auxiliary/thrift/services/Receiver.java
/*
 *  This file is part of the auxiliaries of VIB (Virtual Interactive Behaviour).
 */
package vib.auxiliary.thrift.services;

import vib.auxiliary.thrift.gen_java.Message;
import vib.auxiliary.thrift.gen_java.SimpleCom;
import vib.core.util.log.Logs;
import org.apache.thrift.server.TServer;
import org.apache.thrift.server.TThreadPoolServer;
import org.apache.thrift.transport.TServerSocket;
import org.apache.thrift.transport.TServerTransport;

/**
 *
 * @author Ken Prepin
 */
public abstract class Receiver extends Connector<Receiver> {

    private SimpleComHandler handler;
    private SimpleCom.Processor processor;
    TServerTransport serverTransport;
    TServer server;
    ReceiverThread receiverThread;
    Message message;

    public Receiver() {
        this(DEFAULT_THRIFT_PORT);
    }

    public Receiver(int port) {
        super(DEFAULT_THRIFT_HOST, port);
    }

    @Override
    public void startConnector(Receiver receiver) {
        try {
            handler = new SimpleComHandler(this);
            processor = new SimpleCom.Processor(handler);
            receiverThread = new ReceiverThread(this);
            receiverThread.start();
        } catch (Exception x) {
            x.printStackTrace();
        }
    }
    @Override
    public void stopConnector(Receiver receiver){
       if(receiver.server.isServing()){
           receiver.server.stop();
           receiver.serverTransport.close();
       }
        if(receiver.receiverThread.isAlive()){
            receiver.receiverThread.interrupt();
        }
         receiver.setConnected(false);
    }


    private class ReceiverThread extends Thread {

        final Receiver receiver;

        public ReceiverThread(Receiver receiverToBeStarted) {
            this.setDaemon(true);
            this.receiver = receiverToBeStarted;

        }
        @Override
        public void run() {
            Logs.debug("Try to start on " + receiver.getHost() + " - " + receiver.getPort());
            try {
                receiver.serverTransport = new TServerSocket(getPort());

                // Use this for a monothreaded server ...
                //receiver.server = new TSimpleServer(new Args(serverTransport).processor(processor));

                // ... or use this for a multithreaded server
                receiver.server = new TThreadPoolServer(new TThreadPoolServer.Args(serverTransport).processor(processor));

                System.out.println("Starting the server...");
                setConnected(true);
                server.serve();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public abstract void perform(Message m);
}

File: gretaUnity-master-1/gretaUnity-master/Assets/Scripts/ThriftImpl/services/Receiver.cs
using System;
using UnityEngine;
using System.Threading;
/**
 *
 * @author Ken
 */


using thrift.gen_csharp;
using Thrift;
using Thrift.Protocol;
using Thrift.Transport;
using Thrift.Collections;
using Thrift.Server;

namespace thrift.services
{
	public abstract class Receiver : Connector
	{
 	private SimpleComHandler handler;
    private SimpleCom.Processor processor;
    TServerTransport serverTransport;
    TServer server=null;
    Thread receiverThread;
    Message message;

    public Receiver():this(DEFAULT_THRIFT_PORT) {
        
    }

    public Receiver(int port) : base(DEFAULT_THRIFT_HOST, port){
        
    }
		
    public override void startConnector() {
        try {
            handler = new SimpleComHandler(this);
            processor = new SimpleCom.Processor(handler);
            receiverThread = new Thread(new ThreadStart(startingSimpleServer));
			receiverThread.IsBackground = true; 
            receiverThread.Start();
        } catch (Exception x) {
            Debug.LogError(x);
        }
    }
    
    public override void stopConnector(){
		if(server!=null){
			Debug.Log("Stoping the Receiver...");
           server.Stop();
           serverTransport.Close();
				Debug.Log("Receiver stopped");
       }    
			
		if(receiverThread!=null && receiverThread.IsAlive){
            receiverThread.Abort();
				
        }		   
   
         setConnected(false);
    }
		
    public void startingSimpleServer() {
        
            Debug.Log("Try to start Receiver SimpleServer on " + getHost() + " - " + getPort());
            try {
                serverTransport = new TServerSocket(getPort());
                server = new TSimpleServer(processor,serverTransport);

                // Use this for a multithreaded server
                // TServer server = new TThreadPoolServer(new TThreadPoolServer.Args(serverTransport).processor(processor));

                Debug.Log("Starting the simple server...");
                setConnected(true);
                server.Serve();
            } catch (Exception e) {
                Debug.LogError(e);
            }
        }
    

    public abstract void perform(Message m);
	}
}

